blockchain/event_filter.go

```diff
@@ -0,0 +1,198 @@
+	processedEvents uint64
+}
+
+func (t *ContinuationToken) String() string {
+	return fmt.Sprintf("%d-%d", t.fromBlock, t.processedEvents)
+}
+
+func (t *ContinuationToken) FromString(str string) error {
+	_, err := fmt.Sscanf(str, "%d-%d", &t.fromBlock, &t.processedEvents)
+	return err
+}
+
+type FilteredEvent struct {
+	*core.Event
+	BlockNumber     uint64
+	BlockHash       *felt.Felt
+	TransactionHash *felt.Felt
+}
+
+func (f *EventFilter) Events(cToken *ContinuationToken, chunkSize uint64) ([]*FilteredEvent, *ContinuationToken, error) {
+	var matchedEvents []*FilteredEvent
+
+	filterKeysMap := make(map[felt.Felt]bool, len(f.keys))
+	for _, key := range f.keys {
+		filterKeysMap[*key] = true
+	}
+
+	for curBlock := f.fromBlock; curBlock <= f.toBlock; curBlock++ {
+		// skip the blocks that we previously processed for this request
+		if cToken != nil && curBlock < cToken.fromBlock {
+			continue
+		}
+
+		header, err := blockHeaderByNumber(f.txn, curBlock)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// test `from` only by default if it's not nil because empty filter keys means match all
+		possiblyMatches := true
+		if f.contractAddress != nil {
+			addrBytes := f.contractAddress.Bytes()
+			possiblyMatches = header.EventsBloom.Test(addrBytes[:])
+			// bloom filter says no events from this contract
+			if !possiblyMatches {
+				continue
+			}
+		}
+		for key := range filterKeysMap {
+			keyBytes := key.Bytes()
+
+			// check if block possibly contains the event we are looking for
+			possiblyMatches = header.EventsBloom.Test(keyBytes[:])
+			if possiblyMatches {
+				break
+			}
+		}
+
+		// bloom filter says no events match the filter, skip this block entirely if from is not nil
+		if !possiblyMatches {
+			continue
+		}
+
+		var processedEvents uint64
+		matchedEvents, processedEvents, err = f.appendBlockEvents(matchedEvents, header, filterKeysMap, cToken, chunkSize)
+		if err != nil {
+			if errors.Is(err, errChunkSizeReached) {
+				return matchedEvents, &ContinuationToken{
+					fromBlock:       curBlock,
+					processedEvents: processedEvents,
+				}, nil
+			}
+			return nil, nil, err
+		}
+	}
+	return matchedEvents, nil, nil
+}
+
+func (f *EventFilter) appendBlockEvents(matchedEventsSofar []*FilteredEvent, header *core.Header,
+	keysMap map[felt.Felt]bool, cToken *ContinuationToken, chunkSize uint64,
+) ([]*FilteredEvent, uint64, error) {
+	receipts, err := receiptsByBlockNumber(f.txn, header.Number)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	processedEvents := uint64(0)
+	for _, receipt := range receipts {
+		for _, event := range receipt.Events {
+			// if last request was interrupted mid-block, and we are still processing that block, skip events
+			// that were already processed
+			if cToken != nil && header.Number == cToken.fromBlock && processedEvents < cToken.processedEvents {
+				processedEvents++
+				continue
+			}
+
+			if f.contractAddress != nil && !event.From.Equal(f.contractAddress) {
```

Comment:
What does it mean to have an `f.contractAddress == nil`?

Reply:
https://github.com/NethermindEth/juno/pull/746#discussion_r1197358092